### PR TITLE
Upgrades to Keccak XOR gate and field utils

### PR DESCRIPTION
**Pull request by querolita**
This PR contains code that is common to the other branches for NOT, AND, ROT. It extends utility of XOR gate and adds some field utils.

### CODE REVIEW

File: utils/src/big_bit_ops.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
+//! This module provides a set of functions to perform bit operations on big integers.
+//! In particular, it gives XOR and NOT for BigUint.
+use num_bigint::BigUint;
+use rand::Rng;
+use std::cmp::Ordering;
+
+/// Exclusive or of the bits of two BigUint inputs
+pub fn big_xor(input1: &BigUint, input2: &BigUint) -> BigUint {
```
[the remark this answers is not in the document]
Comment (jspada) [reply utils/src/big_bit_ops.rs:8]: Or maybe it's possible to implement this one.
https://doc.rust-lang.org/std/ops/trait.BitXor.html
or perhaps it's already implemented for `BigUint`.

File: utils/src/field_helpers.rs [hunk 2]
```diff
@@ -41,6 +41,14 @@ pub trait FieldHelpers<F> {
     /// Serialize to bits
     fn to_bits(&self) -> Vec<bool>;
 
+    /// Serialize field element to a BigUint
```
Comment (jspada) [inline utils/src/field_helpers.rs:44]: Not part of this PR.  Update against master to be clean.

File: utils/src/lib.rs [hunk 3]
```diff
@@ -13,8 +14,9 @@ pub mod hasher;
 pub mod math;
 pub mod serialization;
 
+pub use big_bit_ops::{big_bits, big_xor};
 pub use chunked_evaluations::ChunkedEvaluations;
 pub use dense_polynomial::ExtendedDensePolynomial;
 pub use evaluations::ExtendedEvaluations;
-pub use field_helpers::FieldHelpers;
+pub use field_helpers::{FieldFromBig, FieldHelpers};
```
Comment (jspada) [inline utils/src/lib.rs:21]: I think this could just be part of normal FieldHelpers, rather a new trait.

Comment (jspada) [reply utils/src/lib.rs:21]: I think it already exists on master though.. so probably just need to update.

File: kimchi/src/circuits/witness/mod.rs [hunk 4]
```diff
@@ -4,6 +4,7 @@ mod copy_bits_cell;
 mod copy_cell;
 mod copy_shift_cell;
 mod crumb_cell;
```
Comment (jspada) [inline kimchi/src/circuits/witness/mod.rs:6]: This is a misnomer and should be renamed.  A crumb is 2 bits and a nybble is 4.  Maybe this is more acurately `PrefixBitsCell` or something like that?

File: kimchi/src/tests/xor.rs [hunk 5]
```diff
@@ -23,63 +38,128 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
+}
+
+// Returns the all ones BigUint of bits length
+pub(crate) fn all_ones(bits: usize) -> PallasField {
+    PallasField::from(2u128).pow(&[bits as u64]) - PallasField::one()
 }
 
 // Returns a given crumb of 4 bits
-fn xor_crumb(word: u128, crumb: usize) -> u128 {
+pub(crate) fn xor_crumb(word: BigUint, crumb: usize) -> BigUint {
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:70]: Crumb is a misnomer and I think should be renamed to avoid confusion.

**Review by jspada — COMMENTED** [review]
Really nice work.  I didn't spot any bugs or flaws in this gate.  However, it needs an update against master and then I will take another pass over it.

For now, here's a bunch of feedback and suggestions that I hope are useful.

File: kimchi/src/circuits/polynomials/xor.rs [hunk 6]
```diff
@@ -6,7 +6,7 @@ use crate::{
     circuits::{
         argument::{Argument, ArgumentEnv, ArgumentType},
         constraints::ConstraintSystem,
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/polynomials/xor.rs:8]: the file and module are called xor because i want to be general at the library level, but then the gate gives further information about how it works. The only two things I can recall are either `xor.rs` and `Xor16`, I don't think there's any xor16 anywhere

File: kimchi/src/circuits/polynomials/xor.rs [hunk 7]
```diff
@@ -495,32 +464,48 @@ fn init_xor<F: PrimeField>(
...
 pub fn extend_xor_rows<F: PrimeField>(
     witness: &mut [Vec<F>; COLUMNS],
     bits: usize,
     words: (F, F, F),
 ) {
-    let xor_witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero(); num_xors(bits) + 1]);
+    let input1_big = words.0.to_biguint();
+    let input2_big = words.1.to_biguint();
+    let output_big = words.2.to_biguint();
+    if bits < big_bits(&input1_big) || bits < big_bits(&input2_big) || bits < big_bits(&output_big)
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/polynomials/xor.rs:477]: because that gives 0 bits when the input is 0 instead of 1

File: kimchi/src/circuits/witness/sum_copy_bits_cell.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+use ark_ff::Field;
+use o1_utils::FieldHelpers;
+
+use super::{variables::Variables, WitnessCell};
+use crate::circuits::polynomial::COLUMNS;
+
+/// Witness cell copied from bits of another witness cell
+pub struct SumCopyBitsCell<F: Field> {
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/witness/sum_copy_bits_cell.rs:8]: This is because this is not exactly doing what VariableBitsCell does. It is a VariableBitsCell to which you add the value stored in sum. I guess I can replace this logic within the variable map though, I will give it a try.

[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/witness/sum_copy_bits_cell.rs:8]: Worked

File: kimchi/src/prover.rs [hunk 9]
```diff
@@ -632,12 +633,7 @@ where
             }
 
             if let Some(selector) = index.cs.foreign_field_add_selector_poly.as_ref() {
-                index_evals.extend(
-                    foreign_field_add::gadget::circuit_gates()
-                        .iter()
-                        .enumerate()
-                        .map(|(_, gate_type)| (*gate_type, &selector.eval8)),
-                );
+                index_evals.insert(GateType::ForeignFieldAdd, &selector.eval8);
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/prover.rs]: it is indeed intended to be here, the name of this PR is not super descriptive, should be called: "common changes among ethereum primitives pr's" perhaps...

File: kimchi/src/tests/foreign_field_add.rs [hunk 10]
```diff
@@ -424,7 +420,7 @@ fn test_one_minus_one_plus_one() {
 fn test_minus_minus() {
     let foreign_modulus = BigUint::from_bytes_be(SECP256K1_MOD);
     let neg_one_for = ForeignElement::<PallasField, 3>::from_be(ONE).neg(&foreign_modulus);
-    let neg_one = neg_one_for.to_big().to_bytes_be();
+    let neg_one = neg_one_for.to_biguint().to_bytes_be();
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/foreign_field_add.rs]: fixed when merged

File: kimchi/src/tests/xor.rs [hunk 11]
```diff
@@ -23,63 +38,128 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
+) -> [Vec<PallasField>; COLUMNS] {
+    let rng = &mut StdRng::from_seed(RNG_SEED);
+    // Initalize inputs
+    // If some input was given then use that one, otherwise generate a random one with the given bits
+    let input1 = initialize(in1, bits, rng);
+    let input2 = initialize(in2, bits, rng);
+
+    // If user specified a concrete number of bits, use that (if they are sufficient to hold both inputs)
+    // Otherwise, use the max number of bits required to hold both inputs (if only one, the other is zero)
+    let bits1 = big_bits(&input1.to_biguint());
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/xor.rs:127]: not really, as bits() returns 0 if input is 0

File: utils/src/big_bit_ops.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/big_bit_ops.rs:8]: Mmm I like the idea, I will try

[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/big_bit_ops.rs]: Okay, I will have a trait for all of these, but the one inside the std library requires to clone the self because you cannot copy biguint, thus made a new one

File: kimchi/src/circuits/witness/mod.rs (hunk 4 above)
Comment (querolita) [reply kimchi/src/circuits/witness/mod.rs:6]: oh, forgot the nybble name

File: kimchi/src/circuits/polynomials/xor.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/circuits/polynomials/xor.rs:8]: Sounds good

File: kimchi/src/circuits/polynomials/xor.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/circuits/polynomials/xor.rs:484]: whaaa?  Weird

File: kimchi/src/tests/xor.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/tests/xor.rs:147]: seems broken then

File: kimchi/src/tests/xor.rs [hunk 12]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
+// Generates a random field element of up to a given number of bits
+pub(crate) fn random_field<G: KimchiCurve>(bits: usize, rng: &mut StdRng) -> G::ScalarField {
+    G::ScalarField::from_biguint(
+        rng.gen_biguint_range(&BigUint::from(0u8), &BigUint::from(2u8).pow(bits as u32)),
+    )
+    .unwrap()
+}
+
+// Initialize a random input
+pub(crate) fn initialize<G: KimchiCurve>(
```
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/tests/xor.rs:71]: I think it would be better making a `rand()` to `FieldHelpers` and just going `F::rand()`

File: kimchi/src/tests/xor.rs [hunk 13]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
+    witness: &[Vec<G::ScalarField>; COLUMNS],
+    bits: usize,
+    input1: G::ScalarField,
+    input2: G::ScalarField,
+    not: bool,
+) {
+    let input1 = input1.to_biguint();
+    let input2 = input2.to_biguint();
+    let ini_row = if not == XOR { 0 } else { 1 };
+    for x in 0..xor::num_xors(bits) {
```
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/tests/xor.rs:106]: Maybe a more descriptive variable name

File: kimchi/src/tests/xor.rs [hunk 14]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
-            .collect::<Vec<u128>>();
-        for crumb in 0..4 {
+            .map(|i| xor_nybble(input2.clone(), i + 4 * x))
+            .collect::<Vec<BigUint>>();
+        for nybble in 0..4 {
             assert_eq!(
-                witness[11 + crumb][xor],
-                PallasField::from(in1[crumb] ^ in2[crumb])
+                witness[11 + nybble][x + ini_row],
+                G::ScalarField::from(BigUint::bitxor(&in1[nybble], &in2[nybble]))
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:116]: I'm confused about what `BigUint::bitxor` is doing because the output should be a `BigUint` while `ScalarField::from` should not expect a `BigUint` unless it was `ScalarField::from_biguint`

File: kimchi/src/circuits/polynomials/xor.rs [hunk 15]
```diff
@@ -501,32 +471,48 @@ fn init_xor<F: PrimeField>(
...
```
Comment (jspada) [inline kimchi/src/circuits/polynomials/xor.rs:484]: I think you should definitely implement `big_bits` as a `BigUnitHelpers` trait instead, so the code here will look like

```rust
if bits < input1_big.bit_len() || bits < input2_big.bit_len() || bits < output_big.bit_len()
```

File: kimchi/src/tests/xor.rs [hunk 16]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
-// Manually checks the XOR of each crumb in the witness
-fn check_xor(witness: &[Vec<PallasField>; COLUMNS], bits: usize, input1: u128, input2: u128) {
-    for xor in 0..xor::num_xors(bits) {
+// Returns a given nybble of 4 bits
+pub(crate) fn xor_nybble(word: BigUint, nybble: usize) -> BigUint {
+    (word >> (4 * nybble)) % 2u128.pow(4)
+}
+
+// Manually checks the XOR of each nybble in the witness
+pub(crate) fn check_xor<G: KimchiCurve>(
```
Comment (jspada) [inline kimchi/src/tests/xor.rs:96]: Why is this pub(crate) I don't code outside of tests should be using tests code.  The other way around is OK though, so maybe move this to `xor.rs`?

File: utils/src/big_bit_ops.rs [hunk 17]
```diff
@@ -0,0 +1,114 @@
+//! This module provides a set of functions to perform bit operations on big integers.
+//! In particular, it gives XOR and NOT for BigUint.
+use num_bigint::BigUint;
+use rand::Rng;
+use std::cmp::Ordering;
+
+/// Exclusive or of the bits of two inputs
+pub trait BitXor<Rhs = Self> {
+    /// Exclusive or of the bits of two BigUint inputs
```
Comment (jspada) [inline utils/src/big_bit_ops.rs:9]: nit: `// XOR BigUints`

File: utils/src/big_bit_ops.rs [hunk 18]
```diff
@@ -0,0 +1,114 @@
+//! This module provides a set of functions to perform bit operations on big integers.
```
Comment (jspada) [inline utils/src/big_bit_ops.rs:1]: Suggest creating `biguint_helpers.rs` containing all of this implemented as a `BigUintHelpers` trait instead

File: kimchi/src/circuits/polynomials/xor.rs [hunk 19]
```diff
@@ -447,9 +417,9 @@ fn layout<F: PrimeField>(curr_row: usize, bits: usize) -> Vec<[Box<dyn WitnessCe
 
 fn xor_row<F: PrimeField>(crumb: usize, curr_row: usize) -> [Box<dyn WitnessCell<F>>; COLUMNS] {
     [
-        CrumbCell::create("in1", crumb),
-        CrumbCell::create("in2", crumb),
-        CrumbCell::create("out", crumb),
+        NybbleCell::create("in1", crumb),
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/polynomials/xor.rs:420]: Oh, just because I forgot to rename those

File: utils/src/big_bit_ops.rs [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+        .collect::<Vec<u8>>();
+    if extra > 0 {
+        big.push(rand::thread_rng().gen_range(0..2u8.pow(extra as u32)));
+    }
+    BigUint::from_bytes_le(&big)
+}
+
+// Returns a BigUint as a Vec<u8> padded with zeros to a certain number of bytes
+// Panics if bytes < input.len()
+fn vectorize(input: &BigUint, bytes: usize) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/big_bit_ops.rs:56]: oh but this is a private function, not used anywhere else

[the remark this answers is not in the document]
Comment (querolita) [reply utils/src/big_bit_ops.rs:56]: will rename it though

File: kimchi/src/circuits/witness/nybble_cell.rs [hunk 21]
```diff
@@ -0,0 +1,26 @@
+use super::{variables::Variables, WitnessCell};
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/circuits/witness/nybble_cell.rs:1]: Then I will modify VariableBitsCell to allow for an Option end parameter, such that if it is None then it takes all bits till the end (meaning [start..]).

File: kimchi/src/tests/xor.rs [hunk 22]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
-fn test_xor(in1: u128, in2: u128, bits: usize) -> [Vec<PallasField>; COLUMNS] {
-    let cs = create_test_constraint_system(bits);
-    let witness = xor::create(in1, in2, bits);
-    for row in 0..xor::num_xors(bits) + 1 {
-        assert_eq!(
-            cs.gates[row].verify_witness::<Vesta>(row, &witness, &cs, &witness[0][0..cs.public]),
-            Ok(())
-        );
+// Generates a random field element of up to a given number of bits
+pub(crate) fn random_field<G: KimchiCurve>(bits: usize, rng: &mut StdRng) -> G::ScalarField {
```
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/xor.rs:63]: Aaah.... I like your ideas! I will

File: kimchi/src/tests/xor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/xor.rs:71]: similar, because this will be useful for the other PRs, which also use it

File: kimchi/src/tests/xor.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (querolita) [reply kimchi/src/tests/xor.rs:106]: x for each of the Xors, but okay

File: kimchi/src/tests/xor.rs (hunk 14 above)
Comment (querolita) [reply kimchi/src/tests/xor.rs:116]: it just returns the bitwise xor of two biguint inputs, and then i transform it to field format so i can compare with what's in the witness, but I will write it more readable

File: kimchi/src/circuits/polynomials/xor.rs (hunk 15 above)
Comment (querolita) [reply kimchi/src/circuits/polynomials/xor.rs:484]: done!

File: kimchi/src/tests/xor.rs (hunk 16 above)
Comment (querolita) [reply kimchi/src/tests/xor.rs:96]: Oh, but I will use this in AND, and NOT, that's why it is pub(crate) here. This is just one of the ugly parts of having so many open PRs, it is hard to get the general idea of everything when it is so partitioned...

File: kimchi/src/tests/xor.rs [hunk 23]
```diff
@@ -31,58 +59,145 @@ fn create_test_constraint_system(bits: usize) -> ConstraintSystem<Fp> {
...
+        random_field::<G>(bits, rng)
     }
-    witness
 }
 
-// Returns a given crumb of 4 bits
-fn xor_crumb(word: u128, crumb: usize) -> u128 {
-    (word >> (4 * crumb)) % 2u128.pow(4)
+// Returns the all ones BigUint of bits length
+pub(crate) fn all_ones<G: KimchiCurve>(bits: usize) -> G::ScalarField {
```
[the remark this answers is not in the document]
Comment (jspada) [reply kimchi/src/tests/xor.rs:63]: Just for tests?

File: kimchi/src/tests/xor.rs (hunk 16 above)
Comment (jspada) [reply kimchi/src/tests/xor.rs:73]: I see, yeah!

**Review by jspada — APPROVED** [review]
Looks good to me -- approved.  Bravissimo!
